src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java

```diff
@@ -150,6 +162,118 @@ public ControlledGcpResourceApiController(
+            resource,
+            body.getAiNotebookInstance(),
+            privateRoles,
+            body.getJobControl(),
+            ControllerUtils.getAsyncResultEndpoint(
+                request, body.getJobControl().getId(), "create-result"),
+            userRequest);
+
+    ApiCreatedControlledGcpAiNotebookInstanceResult result =
+        fetchNotebookInstanceResult(jobId, userRequest);
+    return new ResponseEntity<>(result, HttpStatus.valueOf(result.getJobReport().getStatusCode()));
+  }
+
+  @Override
+  public ResponseEntity<ApiCreatedControlledGcpAiNotebookInstanceResult>
+      getCreateAiNotebookInstanceResult(UUID workspaceId, String jobId) {
+    AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
+    ApiCreatedControlledGcpAiNotebookInstanceResult result =
+        fetchNotebookInstanceResult(jobId, userRequest);
+    return new ResponseEntity<>(result, HttpStatus.valueOf(result.getJobReport().getStatusCode()));
+  }
+
+  private ApiCreatedControlledGcpAiNotebookInstanceResult fetchNotebookInstanceResult(
+      String jobId, AuthenticatedUserRequest userRequest) {
+    AsyncJobResult<ControlledAiNotebookInstanceResource> jobResult =
+        jobService.retrieveAsyncJobResult(
+            jobId, ControlledAiNotebookInstanceResource.class, userRequest);
+
+    ApiGcpAiNotebookInstanceResource apiResource = null;
+    if (jobResult.getJobReport().getStatus().equals(ApiJobReport.StatusEnum.SUCCEEDED)) {
+      ControlledAiNotebookInstanceResource resource = jobResult.getResult();
+      String workspaceProjectId = workspaceService.getRequiredGcpProject(resource.getWorkspaceId());
+      apiResource = resource.toApiResource(workspaceProjectId);
+    }
+    return new ApiCreatedControlledGcpAiNotebookInstanceResult()
+        .jobReport(jobResult.getJobReport())
+        .errorReport(jobResult.getApiErrorReport())
+        .aiNotebookInstance(apiResource);
+  }
+
+  @Override
+  public ResponseEntity<ApiGcpAiNotebookInstanceResource> getAiNotebookInstance(
+      UUID workspaceId, UUID resourceId) {
+    AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
+    ControlledResource controlledResource =
+        controlledResourceService.getControlledResource(workspaceId, resourceId, userRequest);
+    ApiGcpAiNotebookInstanceResource response =
+        controlledResource
+            .castToAiNotebookInstanceResource()
+            .toApiResource(workspaceService.getRequiredGcpProject(workspaceId));
+    return new ResponseEntity<>(response, HttpStatus.OK);
+  }
+
+  /**
+   * Extract a list of ControlledResourceIamRoles from the common fields of a controlled resource
+   * request body, and validate that it's shaped appropriately for the specified AccessScopeType.
+   *
+   * <p>Shared access resources must not specify private resource roles. Private access resources
+   * must specify at least one private resource role.
+   */
+  private static List<ControlledResourceIamRole> getAndValidatePrivateRoles(
```

Comment:
This looks familiar. Is this a function we should move into a utility?

Reply:
Zach and I were racing PRS. Using the function he added in his BQ PR. (I had meant to finish pulling this out of `createBucket`).

Reply:
The similarity is probably with #307, which also pulled this function out of `createBucket`. there's going to be a small merge conflict here, but I think this controller is approximately the right place for it. maybe we move it to a utility once we support non-GCP controlled resources?

Reply:
Thanks, just merged and updated, PTAL